[UBSAN] add null and alignment checks for aggregates

vasu-the-sharma opened the pull request:
## Description
This PR adds sanitizer checks for null pointers and misalignment during aggregate copy operations in EmitAggregateCopy.

Fixes https://github.com/llvm/llvm-project/issues/61466.

## Current State
When copying aggregates (structs), the compiler does not validate that source and destination pointers are non-null or properly aligned when sanitizers are enabled. This allows undefined behavior from invalid pointers to go undetected at runtime.

## Solution
When null or alignment sanitizers are active, this change:

- Retrieves raw pointer addresses from both source and destination LValues
- Performs EmitTypeCheck on both pointers to verify they are non-null and meet alignment requirements
- Records these checks in the SanitizerSet for runtime emission

This ensures memory safety violations are caught early, preventing crashes and corruption from invalid aggregate copies.

## Testing
Added comprehensive test cases covering:
- Null source pointer detection
- Null destination pointer detection
- Misaligned source pointer detection
- Misaligned destination pointer detection
- Normal copy operations (no false positives)

## Screenshots
<img width="2842" height="2040" alt="image (4)" src="https://github.com/user-attachments/assets/b5c8e85a-cb4f-4251-be6b-c216adff08e9" />
<img width="2862" height="576" alt="image (5)" src="https://github.com/user-attachments/assets/d7a20d04-2fd4-4628-8257-d872a48a5d5b" />

Commit:
add null and aligment checks for aggregates

vasu-the-sharma committed:
Merge branch 'main' into vasu/add-null-and-alignment-checks-for-aggregates

tonykuttai commented on clang/lib/CodeGen/CGExprAgg.cpp:
```diff
@@ -2249,6 +2249,21 @@ void CodeGenFunction::EmitAggregateCopy(LValue Dest, LValue Src, QualType Ty,
                                         bool isVolatile) {
   assert(!Ty->isAnyComplexType() && "Shouldn't happen for complex");
 
+  if (SanOpts.hasOneOf(SanitizerKind::Null | SanitizerKind::Alignment)) {
```
Can you add comment block here to explain
- why we need this `if block`
- what is it doing.?

tonykuttai reviewed (changes requested):
Please add a test support this change

Commit:
Add null and alignment checks for aggregate copy operation

vasu-the-sharma committed:
Merge branch 'main' into vasu/add-null-and-alignment-checks-for-aggregates

hubert-reinterpretcast reviewed (commented):
I believe there should be a Clang CodeGen test for this.

Commit:
add codegen test

vasu-the-sharma commented:
ping

hubert-reinterpretcast commented on clang/lib/CodeGen/CGExprAgg.cpp:
```diff
@@ -2249,6 +2249,24 @@ void CodeGenFunction::EmitAggregateCopy(LValue Dest, LValue Src, QualType Ty,
+    Address DestAddr = Dest.getAddress();
+
+    // Check source pointer for null and alignment violations
+    EmitTypeCheck(TCK_Load, SourceLocation(),
```
`EmitCheckedLvalue` will consider the expression to be accessed (and therefore flag past-the-end array elements).

Consider:
```cpp
#if !SCALAR
typedef struct A { int x; } A;
#else
typedef int A;
#endif
A arr[1];
A *ap = arr;

int main(void) {
  A a = { 0 };
  arr[1] = a;
}
```

It looks like `EmitAggregateCopy` may be too "low-level" for all the necessary information to be available.

hubert-reinterpretcast replied:
Consider changes like those from https://github.com/vasu-the-sharma/llvm-project/pull/1.
I have not updated and checked the tests in detail though (and some additional test coverage will probably be needed).

hubert-reinterpretcast replied:
For ease of review, it may make sense to pre-commit the test before this PR (so that the change in this PR will simply be the addition of the sanitizer checks).

hubert-reinterpretcast committed:
Call `EmitCheckedLValue` from additional `AggExprEmitter` functions

... instead of calling `EmitTypeCheck` directly from `EmitAggregateCopy`
(at which point contextual information, including the source location,
is already lost).

hubert-reinterpretcast commented on clang/test/CodeGen/ubsan-aggregate-null-align.cpp:
```diff
@@ -0,0 +1,134 @@
+  int z;
+};
+
+// Stack-to-stack copies are optimized away (compiler knows they're valid)
```
This needs `-NOT` check lines to actually test that the cases are optimized away; however, I don't think this is something we need to test in this PR because there should already be tests covering the underlying internal Clang APIs that we are calling. On top of the functional verification (that various specific checks are implemented), we only need enough to demonstrate that we are probably passing the right arguments to the APIs.

hubert-reinterpretcast commented on clang/test/CodeGen/ubsan-aggregate-null-align.cpp:
```diff
@@ -0,0 +1,134 @@
+
+// CHECK-UBSAN-LABEL: define {{.*}}void @_Z18test_normal_structv()
+// CHECK-NO-UBSAN-LABEL: define {{.*}}void @_Z18test_normal_structv()
+void test_normal_struct() {
```
Generally the same comment as above. The testing associated with this function as written does not really cover much (and maybe the test is not helpful).

hubert-reinterpretcast commented on clang/test/CodeGen/ubsan-aggregate-null-align.cpp:
```diff
@@ -0,0 +1,134 @@
+  // CHECK-UBSAN: call void @__ubsan_handle_type_mismatch_v1_abort
+  // CHECK-UBSAN: unreachable
+
+  // CHECK-UBSAN: cont{{.*}}:
```
The label can also be made a `FileCheck` variable.

vasu-the-sharma replied:
Good point! I've updated the CHECK patterns to use FileCheck variables for capturing and 
matching labels dynamically.

hubert-reinterpretcast commented on clang/test/CodeGen/ubsan-aggregate-null-align.cpp:
```diff
@@ -0,0 +1,134 @@
+  *dest = *src;
+}
+
+// Array copies also need checks for non-constant indices
```
The test does not use non-constant indices though. Please modify to perform a single access with a non-constant index (instead of performing multiple accesses with constant index values).

hubert-reinterpretcast committed:
Update test to compile as either C or C++

hubert-reinterpretcast commented on clang/test/CodeGen/ubsan-aggregate-null-align.cpp:
```diff
@@ -0,0 +1,134 @@
+// RUN: %clang_cc1 -triple x86_64-linux-gnu -fsanitize=alignment,null \
+// RUN:   -emit-llvm %s -o - | FileCheck %s --check-prefix=CHECK-UBSAN
+// RUN: %clang_cc1 -triple x86_64-linux-gnu -emit-llvm %s -o - \
+// RUN:   | FileCheck %s --check-prefix=CHECK-NO-UBSAN
```
Suggestion: Remove the cases with no sanitizer enabled. Coverage almost certainly exists already.

hubert-reinterpretcast committed:
Use EmitCheckedLValue for C++ trivial assignment operator calls

hubert-reinterpretcast committed:
(partial) Use EmitCheckedLValue for C++ trivial copy/move ctor calls

vitalybuka commented on clang/lib/CodeGen/CGExprAgg.cpp:
```diff
@@ -2249,6 +2249,22 @@ void CodeGenFunction::EmitAggregateCopy(LValue Dest, LValue Src, QualType Ty,
+  // Check source pointer for null and alignment violations
+  EmitTypeCheck(TCK_Load, SourceLocation(),
+                SrcAddr.emitRawPointer(*this), Ty, SrcAddr.getAlignment(),
+                SanitizerSet());
```
SanitizerSet() is default

vitalybuka reviewed (changes requested):
We need a real SourceLocation().

I applied the patch and see:

`test.so+0xbc38): runtime error store to misaligned address`

but for existing alignment checks I see:
`test.cc:221:39: runtime error: load of misaligned`

I guess we need to forward `getExprLoc` from callers.

vitalybuka commented:
> > We need a real SourceLocation().
> 
> The approach from [vasu-the-sharma#1](https://github.com/vasu-the-sharma/llvm-project/pull/1) will get us real source locations (but does not cover all cases where `EmitAggregateCopy` is called).

For UBSAN we need to keep the location. It's supposed to be usable without debug info. Without location it's hard to act on such reports. 

However, how hard to pass it into EmitAggregateCopy?
Maybe we can fork EmitAggregateCopy -> EmitAggregateCopyCurrentOne + EmitAggregateCopyWithSourceLocation 
and incrementally transition from first to another?

hubert-reinterpretcast commented:
> > > We need a real SourceLocation().
> > 
> > 
> > The approach from [vasu-the-sharma#1](https://github.com/vasu-the-sharma/llvm-project/pull/1) will get us real source locations (but does not cover all cases where `EmitAggregateCopy` is called).
> 
> For UBSAN we need to keep the location. It's supposed to be usable without debug info. Without location it's hard to act on such reports.
> 
> However, how hard to pass it into EmitAggregateCopy? Maybe we can fork EmitAggregateCopy -> EmitAggregateCopyCurrentOne + EmitAggregateCopyWithSourceLocation and incrementally transition from first to another?

The source location is not the only issue. As noted in https://github.com/llvm/llvm-project/pull/164548#discussion_r2529214150, if we use `EmitCheckedLValue` while we still have the source expression, we get better checking for array subscript operations.

Additionally, the approach in https://github.com/vasu-the-sharma/llvm-project/pull/1 also covers some cases where we don't (always) call `EmitAggregateCopy` such as in https://github.com/vasu-the-sharma/llvm-project/blob/bc7f979bcb205b82d43d99fd940889f3d8801ca4/clang/lib/CodeGen/CGExprAgg.cpp#L1331-L1341.

TL;DR: The issue is not necessarily with `EmitAggregateCopy` not generating UBSAN checks. The case can be made that the issue is that the `LValue` arguments did not come from calls to `EmitCheckedLValue`.

hubert-reinterpretcast commented:
> > However, how hard to pass it into EmitAggregateCopy? Maybe we can fork EmitAggregateCopy -> EmitAggregateCopyCurrentOne + EmitAggregateCopyWithSourceLocation and incrementally transition from first to another?
> 
> The source location is not the only issue. As noted in [#164548 (comment)](https://github.com/llvm/llvm-project/pull/164548#discussion_r2529214150), if we use `EmitCheckedLValue` while we still have the source expression, we get better checking for array subscript operations.

@vitalybuka, ping.

IMO, adding checks in `EmitAggregateCopy` is too late. Adding calls to `EmitCheckedLValue` earlier in the AST processing that eventually calls `EmitAggregateCopy` is more comprehensive. A test-driven approach that identifies the AST processing paths for common cases strikes a good development cost/user benefit ratio. When further cases are identified, they can be addressed individually without a large upfront refactoring.

vitalybuka commented:
> > > However, how hard to pass it into EmitAggregateCopy? Maybe we can fork EmitAggregateCopy -> EmitAggregateCopyCurrentOne + EmitAggregateCopyWithSourceLocation and incrementally transition from first to another?
> > 
> > 
> > The source location is not the only issue. As noted in [#164548 (comment)](https://github.com/llvm/llvm-project/pull/164548#discussion_r2529214150), if we use `EmitCheckedLValue` while we still have the source expression, we get better checking for array subscript operations.
> 
> @vitalybuka, ping.
> 
> IMO, adding checks in `EmitAggregateCopy` is too late. Adding calls to `EmitCheckedLValue` earlier in the AST processing that eventually calls `EmitAggregateCopy` is more comprehensive. A test-driven approach that identifies the AST processing paths for common cases strikes a good development cost/user benefit ratio. When further cases are identified, they can be addressed individually without a large upfront refactoring.

LGTM

vasu-the-sharma committed:
Merge pull request #1 from hubert-reinterpretcast/hstong/UseEmitCheckedLValueForAggregateCopy

WIP: Use EmitCheckedLValue for aggregate copy

vasu-the-sharma committed:
Merge branch 'main' into vasu/add-null-and-alignment-checks-for-aggregates

vasu-the-sharma committed:
Merge branch 'main' into vasu/add-null-and-alignment-checks-for-aggregates

Commit:
use EmitCheckedLValue at higher levels rather than adding checks in EmitAggregateCopy
